### PR TITLE
Redesign HTTP listener/client SecureSocket API

### PR DESCRIPTION
**Pull request by ldclakmal**
## Purpose
This PR redesign the HTTP listener/client `SecureSocket` API for SwanLake release. The updated API would be as follows:

**Listener**
```ballerina
public type ListenerSecureSocket record {|
   crypto:KeyStore|CertKey key;
   record {|
       VerifyClient verifyClient = REQUIRE;
       crypto:TrustStore|string cert;
   |} mutualSsl?;
   record {|
       Protocol name;
       string[] versions = [];
   |} protocol?;
   record {|
       CertValidationType type = OCSP_STAPLING;
       int cacheSize;
       decimal cacheValidityPeriod;
   |} certValidation?;
   string[] ciphers = [];
   boolean shareSession = true;
   decimal handshakeTimeout?;
   decimal sessionTimeout?;
|};
 
public type CertKey record {|
   string certFile;
   string keyFile;
   string keyPassword?;
|};
 
public enum VerifyClient {
   REQUIRE,
   OPTIONAL
}
 
public enum Protocol {
   SSL,
   TLS,
   DTLS
}
 
public enum CertValidationType {
   OCSP_CRL,
   OCSP_STAPLING
}
```

**Client**
```ballerina
public type ClientSecureSocket record {|
   boolean enable = true;
   crypto:TrustStore|string cert?;
   crypto:KeyStore|CertKey key?;
   record {|
       Protocol name;
       string[] versions = [];
   |} protocol?;
   record {|
       CertValidationType type = OCSP_STAPLING;
       int cacheSize;
       decimal cacheValidityPeriod;
   |} certValidation?;
   string[] ciphers?;
   boolean verifyHostName = true;
   boolean shareSession = true;
   decimal handshakeTimeout?;
   decimal sessionTimeout?;
|};
 
public type CertKey record {|
   string certFile;
   string keyFile;
   string keyPassword?;
|};
 
public enum Protocol {
   SSL,
   TLS,
   DTLS
}
 
public enum CertValidationType {
   OCSP_CRL,
   OCSP_STAPLING
}
```

Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/917

Related to https://github.com/ballerina-platform/ballerina-standard-library/issues/584

### CODE REVIEW

File: http-ballerina/http_client_endpoint.bal [hunk 1]
```diff
@@ -425,37 +425,34 @@ public type RetryConfig record {|
...
-# + keyStore - Configurations associated with KeyStore
-# + certFile - A file containing the certificate of the client
-# + keyFile - A file containing the private key of the client
-# + keyPassword - Password of the private key if it is encrypted
-# + trustedCertFile - A file containing a list of certificates or a single certificate that the client trusts
+# + cert - Configurations associated with `crypto:TrustStore` or single certificate file that the client trusts
+# + key - Configurations associated with `crypto:KeyStore` or combination of certificate and private key of the client
 # + protocol - SSL/TLS protocol related options
-# + certValidation - Certificate validation against CRL or OCSP related options
+# + certValidation - Certificate validation against CRL_OCSP, OCSP_STAPLING related options
```
Comment (Bhashinee) [inline http-ballerina/http_client_endpoint.bal:431]:
```suggestion
# + certValidation - Certificate validation against OCSP_CRL, OCSP_STAPLING related options
```
Let's switch the CRL and OCSP.

File: http-ballerina/http_service_endpoint.bal [hunk 2]
```diff
@@ -187,42 +187,70 @@ public type RequestLimitConfigs record {|
...
 # + protocol - SSL/TLS protocol related options
-# + certValidation - Certificate validation against CRL or OCSP related options
-# + ciphers - List of ciphers to be used (e.g.: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
-#             TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA)
-# + sslVerifyClient - The type of client certificate verification. (e.g.: "require" or "optional")
-# + shareSession - Enable/disable new SSL session creation
-# + handshakeTimeoutInSeconds - SSL handshake time out
-# + sessionTimeoutInSeconds - SSL session time out
-# + ocspStapling - Enable/disable OCSP stapling
+# + certValidation - Certificate validation against CRL_OCSP, OCSP_STAPLING related options
```
Comment (ldclakmal) [inline http-ballerina/http_service_endpoint.bal:193]:
```suggestion
# + certValidation - Certificate validation against OCSP_CRL, OCSP_STAPLING related options
```

File: http-native/src/main/java/org/ballerinalang/net/http/HttpUtil.java [hunk 3]
```diff
@@ -1621,139 +1547,183 @@ private static String getServerName() {
...
+    private static void evaluateCertValidationField(BMap<BString, Object> certValidation,
+                                                    SslConfiguration sslConfiguration) {
+        String type = certValidation.getStringValue(SECURESOCKET_CONFIG_CERT_VALIDATION_TYPE).getValue();
+        if (type.equals(SECURESOCKET_CONFIG_CERT_VALIDATION_TYPE_OCSP_STAPLING.getValue())) {
+            sslConfiguration.setOcspStaplingEnabled(true);
+        }
+        long cacheSize = certValidation.getIntValue(SECURESOCKET_CONFIG_CERT_VALIDATION_CACHE_SIZE).intValue();
+        long cacheValidityPeriod = ((BDecimal) certValidation.get(
+                SECURESOCKET_CONFIG_CERT_VALIDATION_CACHE_VALIDITY_PERIOD)).intValue();
+        sslConfiguration.setValidateCertEnabled(true);
```
Comment (Bhashinee) [inline http-native/src/main/java/org/ballerinalang/net/http/HttpUtil.java:1685]: Shall we move the above line to the else part of `if (type.equals(SECURESOCKET_CONFIG_CERT_VALIDATION_TYPE_OCSP_STAPLING.getValue())) {`

?

Comment (ldclakmal) [reply http-native/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: Fixed with https://github.com/ballerina-platform/module-ballerina-http/pull/255/commits/1a7dc49ebf025aeddaec010e193c4e184353d2bb

**Review by Bhashinee — APPROVED** [review]

**Pull request merged**